src/libraries/Common/src/System/Diagnostics/TraceListenerHelpers.cs

```diff
@@ -13,5 +13,29 @@ internal static int GetThreadId()
         {
             return Environment.CurrentManagedThreadId;
         }
+
+        internal static int GetProcessId()
+        {
+            int id;
+            using (var process = Process.GetCurrentProcess())
```

Comment:
This might regress perf in this case.  Before there was just a single PInvoke, now there is a bit more (allocation, a few more method calls).  It also removed caching on Windows. cc @safern who added this originally.

Reply:
Likely more than a little.  Even without caching, the `Process.GetCurrentProcess().Id` approach is ~50x slower in throughput (and allocates ~250 bytes) than just making the P/Invoke call (which allocates nothing).  Add the caching, and the difference is even larger.

Without a much more convincing argument about why this is important to take, we shouldn't.

Reply:
I wonder if we should have CoreLib or Proceess expose a public GetCurrentProcessId() that does the right thing for the platform.  I took a quick scan and see GetPid called in 4 places, including CoreLib where it is [cached](https://github.com/dotnet/runtime/blob/bd6cbe3642f51d70839912a6a666e5de747ad581/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L2972) despite forking.

Reply:
An Environment.ProcessId property would be reasonable, IMO.

Reply:
We can add caching here, but i am not sure about its effect on the unix code path
```
// Whereas the Win32 implementation caches the GetProcessId result, the Unix
// implementation doesn't so as to avoid problems with fork'd child processes
// ending up returning the same id as the parent.
```

Reply:
That comment seems like it could only be an issue if managed code ran in a fork'd process before exec is called, but if that happened, there are plenty of other things that could fail in much worse ways, e.g. the fork'd process is only going to inherit one thread, which will likely wreak havoc on the GC, cause hangs, etc.  In short, managed code is not fork-safe.  So, while the comment may have been accurate at the time it was written, I don't believe it is now, and we should be able to cache the current pid.

Reply:
Calling `Process.GetCurrentProcess` just to get current process ID is very unfriendly to IL linking. The transitive closure of the `System.Diagnostic.Process` is huge (MBs of code). It is not something that can be fixed by caching.

I agree that the right fix would be to expose light weight current process ID property on e.g. Environment.

Reply:
This already uses Process in the same type to get ProcessName though that appears to be only called in one of the assemblies (the code actually gets removed from TraceSource by the linker in the shared framework).   So this change would regress linked size of TraceSource.

If we wanted to benefit in the TextWriterTraceListener case as well we'd need a method on Environment for ProcessName as well.

Reply:
I opened https://github.com/dotnet/runtime/issues/38388